### PR TITLE
Improve blockchain math precision and safety for U256/I256 conversions

### PR DESCRIPTION
**Pull request by nicolad**
**Fixes precision bugs in blockchain value conversions**

### Problem
- Original implementation converted to `f64` before scaling, causing silent precision loss
- No safety guards for values exceeding `f64`'s exact integer range (2^53-1)
- Adjacent U256 values could collapse to the same f64 result

### Solution
- **Never converts to f64 before scaling** - uses pure integer math first
- **Adds safety guards** - refuses conversion when integer part > 2^53-1
- **Explicit rounding** - caps total significant digits to 15 with proper carry handling
- **Exact string representation** - new `u256_to_decimal_string()` for UI/logging

### Key Changes
- New `convert_u256_to_f64_checked()` with bounds checking and explicit rounding
- New `u256_to_decimal_string()` for exact decimal representation
- Enhanced existing functions to use safe conversion internally
- Comprehensive test suite with 11 new tests covering edge cases and metamorphic properties

### Impact
- **Maintains backward compatibility** - existing API unchanged
- **No new dependencies** - uses only existing `alloy_primitives` and `anyhow`
- **All 20 tests passing** - includes both new safety tests and legacy compatibility tests

Eliminates precision corruption while providing both exact string output and safe f64 conversion

### CODE REVIEW

File: crates/adapters/blockchain/src/math.rs [hunk 1]
```diff
@@ -18,191 +18,340 @@
...
+
+/// Conservative total significant digits we'll aim to keep in f64.
+const MAX_SIG_DIGITS: usize = 15;
+
+/// Maximum decimals s.t. 10^decimals fits in 256 bits.
+/// floor(log10(2^256-1)) = 77
+const MAX_DECIMALS_FIT: u32 = 77;
+
+/// Compute 10^d as U256 (d <= 77).
+fn pow10_u256(d: u32) -> Result<U256> {
```
Comment (filipmacek) [inline crates/adapters/blockchain/src/math.rs]: Just small nitpick: pls use `anyhow::Result` instead of `Result` so that we are separating `anyhow` result from Rust native `Result`. This is standard practice across codebase

File: crates/adapters/blockchain/src/math.rs [hunk 2]
```diff
@@ -18,191 +18,340 @@
...
+    }
+    let mut p = U256::from(1u8);
+    if d == 0 {
+        return Ok(p);
+    }
+    let ten = U256::from(10u8);
+    for _ in 0..d {
+        p *= ten;
+    }
+    Ok(p)
```
Comment (filipmacek) [inline crates/adapters/blockchain/src/math.rs]: The iterative multiplication can be replaced with `U256's` built-in power function for better performance and clarity:
```
fn pow10_u256(d: u32) -> Result<U256> {
    if d > MAX_DECIMALS_FIT {
        bail!("decimals={} exceeds 10^d capacity for U256", d);
    }
    // Safe: 10^77 < 2^256, so this cannot overflow
    Ok(U256::from(10).pow(U256::from(d)))
}
```

File: crates/adapters/blockchain/src/math.rs [hunk 3]
```diff
@@ -18,191 +18,340 @@
...
+    let mut frac_str = frac.to_string();
+    let d = decimals as usize;
+    if frac_str.len() < d {
+        let mut s = String::with_capacity(d);
+        for _ in 0..(d - frac_str.len()) {
+            s.push('0');
+        }
+        s.push_str(&frac_str);
+        frac_str = s;
+    }
```
Comment (filipmacek) [inline crates/adapters/blockchain/src/math.rs]:
```
let frac_str = format!("{:0>width$}", frac, width = decimals as usize);
```
This replaces the manual string building loop and is more idiomatic. And its way easier to read IMO
And can you remove unecesarry  `debug_assert_eq`, length is guaranteed by `format!`  specification

File: crates/adapters/blockchain/src/math.rs [hunk 4]
```diff
@@ -18,191 +18,340 @@
...
+/// Returns an error if the decimals parameter exceeds the maximum supported value.
+pub fn u256_to_decimal_string(amount: U256, decimals: u32) -> Result<String> {
+    let (int_part, mut frac_str) = u256_scaled_parts(amount, decimals)?;
+    if decimals == 0 {
+        return Ok(int_part.to_string());
+    }
+    // Trim trailing zeros in the fractional part; drop '.' if becomes empty.
+    while frac_str.ends_with('0') {
+        frac_str.pop();
+    }
```
Comment (filipmacek) [inline crates/adapters/blockchain/src/math.rs]:
```
frac_str = frac_str.trim_end_matches('0').to_string();
```

File: crates/adapters/blockchain/src/math.rs [hunk 5]
```diff
@@ -18,191 +18,340 @@
...
+    if frac_str.is_empty() {
+        return Ok(int_part_u64 as f64);
+    }
 
-    // Apply sign
-    if is_negative {
-        amount_f64 = -amount_f64;
+    // Remove trailing zeros (no information content).
+    while frac_str.ends_with('0') {
+        frac_str.pop();
```
Comment (filipmacek) [inline crates/adapters/blockchain/src/math.rs]:
```
frac_str = frac_str.trim_end_matches('0').to_string();
```

File: crates/adapters/blockchain/src/math.rs [hunk 6]
```diff
@@ -18,191 +18,340 @@
...
+        } else {
+            (int_part_u64 as f64).log10().floor() as usize + 1
+        }
+    };
+    let keep_frac = if MAX_SIG_DIGITS > int_digits {
+        let budget = MAX_SIG_DIGITS - int_digits;
+        core::cmp::min(budget, frac_str.len())
+    } else {
+        0
+    };
```
Comment (filipmacek) [inline crates/adapters/blockchain/src/math.rs]:
```
    let keep_frac = MAX_SIG_DIGITS.saturating_sub(int_digits).min(frac_str.len());
```

File: crates/adapters/blockchain/src/math.rs [hunk 7]
```diff
@@ -18,191 +18,340 @@
...
-        // 12 decimals
-        let result = convert_i256_to_f64(amount, 12).unwrap();
-        assert_eq!(result, 0.001);
+    fn f64_guard_allows_max_safe_integer() {
+        let amt = u256_dec("9007199254740991"); // 2^53 - 1
+        let v = convert_u256_to_f64_checked(amt, 0).unwrap();
+        assert_eq!(v, 9007199254740991.0);
     }
 
     #[test]
```
Comment (filipmacek) [inline crates/adapters/blockchain/src/math.rs]: pls use `rstest`

File: crates/adapters/blockchain/src/math.rs [hunk 8]
```diff
@@ -18,191 +18,340 @@
...
+                    .ok_or_else(|| anyhow!("integer carry overflow during rounding"))?;
+                if new_int > MAX_SAFE_INT {
+                    bail!("rounded integer {} exceeds f64 exact range.", new_int);
+                }
+                // Update integer part and reset fractional part to all zeros.
+                // Since all zeros add nothing, we can drop them entirely.
+                return Ok(new_int as f64);
+            }
+        }
+    }
```
Comment (filipmacek) [inline crates/adapters/blockchain/src/math.rs]: Would you use the numerical operations here and bytes instead of strings and chars, it is easier to read and better performant. 
So something like 
```
 if frac_str.len() > keep_frac {
        let next_digit = frac_str.as_bytes()[keep_frac];
        frac_str.truncate(keep_frac);

        if next_digit >= b'5' {
            // Round up numerically
            let frac_value = frac_str.parse::<u64>()? + 1;

            // Check for overflow (e.g., 999 + 1 = 1000)
            if frac_value >= 10_u64.pow(keep_frac as u32) {
                // Fraction overflowed: increment integer part
                match int_part_u64.checked_add(1) {
                    Some(new_int) if new_int <= MAX_SAFE_INT => return Ok(new_int as f64),
                    Some(new_int) => bail!("rounded integer {} exceeds f64 exact range.", new_int),
                    None => bail!("integer overflow during rounding"),
                }
            }

            // Format back with leading zeros preserved
            frac_str = format!("{:0>width$}", frac_value, width = keep_frac);
        }
    }
```

**Comment by v9ai** [comment]
change requests applied
